rustfst-ffi/src/fst.rs

```diff
@@ -0,0 +1,359 @@
+        fst.take_output_symbols();
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_final(
+    fst: *mut CFst,
+    state: CStateId,
+    weight: libc::c_float,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.set_final(state, TropicalWeight::new(weight as f32))?;
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_is_final(
+    fst: *const CFst,
+    state: CStateId,
+    is_final: *mut libc::size_t,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst.is_final(state)?;
+        unsafe { *is_final = res as usize }
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_final_weight(
+    fst: *const CFst,
+    state_id: CStateId,
+    final_weight: *mut libc::c_float,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst
+            .final_weight(state_id)?
+            .ok_or_else(|| anyhow!("State '{:?}' is NOT final", state_id))?;
+        unsafe { *final_weight = *res.value() };
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_start(fst: *mut CFst, state: CStateId) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.set_start(state)?;
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_start(fst: *const CFst, state: *mut CStateId) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst
+            .start()
+            .ok_or_else(|| anyhow!("FST has no start state"))?;
+        unsafe { *state = res }
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_equals(
+    fst: *const CFst,
+    other_fst: *const CFst,
+    is_equal: *mut libc::size_t,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let other_fst = get!(CFst, other_fst);
+        let res = fst.eq(other_fst);
+        unsafe { *is_equal = res as usize }
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_draw(
+    fst_ptr: *mut CFst,
+    isyms: *const CSymbolTable,
+    osyms: *const CSymbolTable,
+    fname: *const libc::c_char,
+    title: *const libc::c_char,
+    acceptor: libc::size_t,
+    width: libc::c_float,
+    height: libc::c_float,
+    portrait: libc::size_t,
+    vertical: libc::size_t,
+    ranksep: libc::c_float,
+    nodesep: libc::c_float,
+    fontsize: libc::size_t,
+    show_weight_one: libc::size_t,
+    print_weight: libc::size_t,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst_ptr);
+
+        if !isyms.is_null() {
+            let isymt = get!(CSymbolTable, isyms);
+            fst.set_input_symbols(isymt.clone());
+        }
+
+        if !osyms.is_null() {
+            let osymt = get!(CSymbolTable, osyms);
+            fst.set_output_symbols(osymt.clone());
+        }
+
+        let drawing_config = DrawingConfig {
```

Comment:
What do you think about a dedicated struct for the drawing config ? or the config object exists on python side ?